### PR TITLE
Raise an exception if unable to find a reference for a path and handle symlink references

### LINKED ISSUE #5427: NoMethodError - undefined method `path' for nil:NilClass
# Report
## What did you do?

pod 'wax', :git => 'https://github.com/zrhdev/wax.git', :tag => '1.2.0'

```
/usr/local/bin/pod install --no-repo-update
```
## What did you expected to happen?

Install all pod correctly.
## What happened instead?

The same question of #4887 have not been closed.
### Error

```
NoMethodError - undefined method `path' for nil:NilClass
/Library/Ruby/Gems/2.0.0/gems/xcodeproj-1.0.0/lib/xcodeproj/project/object/native_target.rb:442:in `block in add_file_references'
/Library/Ruby/Gems/2.0.0/gems/xcodeproj-1.0.0/lib/xcodeproj/project/object/native_target.rb:441:in `map'
/Library/Ruby/Gems/2.0.0/gems/xcodeproj-1.0.0/lib/xcodeproj/project/object/native_target.rb:441:in `add_file_references'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.0.0/lib/cocoapods/installer/target_installer/pod_target_installer.rb:94:in `block in add_files_to_build_phases'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.0.0/lib/cocoapods/installer/target_installer/pod_target_installer.rb:75:in `each'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.0.0/lib/cocoapods/installer/target_installer/pod_target_installer.rb:75:in `add_files_to_build_phases'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.0.0/lib/cocoapods/installer/target_installer/pod_target_installer.rb:21:in `block in install!'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.0.0/lib/cocoapods/user_interface.rb:144:in `message'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.0.0/lib/cocoapods/installer/target_installer/pod_target_installer.rb:17:in `install!'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.0.0/lib/cocoapods/installer.rb:621:in `block (2 levels) in install_libraries'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.0.0/lib/cocoapods/installer.rb:619:in `each'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.0.0/lib/cocoapods/installer.rb:619:in `block in install_libraries'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.0.0/lib/cocoapods/user_interface.rb:144:in `message'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.0.0/lib/cocoapods/installer.rb:618:in `install_libraries'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.0.0/lib/cocoapods/installer.rb:170:in `block in generate_pods_project'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.0.0/lib/cocoapods/user_interface.rb:63:in `section'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.0.0/lib/cocoapods/installer.rb:167:in `generate_pods_project'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.0.0/lib/cocoapods/installer.rb:119:in `install!'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.0.0/lib/cocoapods/command/install.rb:37:in `run'
/Library/Ruby/Gems/2.0.0/gems/claide-1.0.0/lib/claide/command.rb:334:in `run'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.0.0/lib/cocoapods/command.rb:50:in `run'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.0.0/bin/pod:55:in `<top (required)>'
/usr/local/bin/pod:22:in `load'
/usr/local/bin/pod:22:in `<main>'
```
### Stack

```
   CocoaPods : 1.0.0
        Ruby : ruby 2.0.0p648 (2015-12-16 revision 53162) [universal.x86_64-darwin15]
    RubyGems : 2.6.4
        Host : Mac OS X 10.11.5 (15F34)
       Xcode : 7.3.1 (7D1014)
         Git : git version 2.7.4 (Apple Git-66)
Ruby lib dir : /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib
Repositories : cms - http://gitlab.365eche.com/Frame/CMSpecs.git @ 7906b0bf8dea0ee498dff7e3d6db67eb1cbe938a
               master - https://github.com/CocoaPods/Specs.git @ c6816cefd545836066f614647ec6a91dca2d57ea
```
### Installation Source

```
Executable Path: /usr/local/bin/pod
```
### Plugins

```
cocoapods-deintegrate : 1.0.0
cocoapods-plugins     : 1.0.0
cocoapods-search      : 1.0.0
cocoapods-stats       : 1.0.0
cocoapods-trunk       : 1.0.0
cocoapods-try         : 1.0.0
```
## Project that demonstrates the issue

pod 'wax', :git => 'https://github.com/zrhdev/wax.git', :tag => '1.2.0'

### LINKED ISSUE #5525: NoMethodError - undefined method `path' for nil:NilClass
- [x] I've read and understood the [*CONTRIBUTING guidelines and have done my best effort to follow](https://github.com/CocoaPods/CocoaPods/blob/master/CONTRIBUTING.md).
# Report
## What did you do?

 Run `pod install`
## What did you expect to happen?

for all pods to be installed

pod 'AWSAutoScaling'
pod 'AWSCloudWatch'
pod 'AWSCognito'
pod 'AWSCognitoIdentityProvider'
pod 'AWSDynamoDB'
pod 'AWSEC2'
pod 'AWSElasticLoadBalancing'
pod 'AWSIoT'
pod 'AWSKinesis'
pod 'AWSLambda'
pod 'AWSMachineLearning'
pod 'AWSMobileAnalytics'
pod 'AWSS3'
pod 'AWSSES'
pod 'AWSSimpleDB'
pod 'AWSSNS'
pod 'AWSSQS'

Install all pod dependencies correctly.
### Error

```
NoMethodError - undefined method `path' for nil:NilClass
/usr/local/Cellar/cocoapods/1.0.1/libexec/gems/xcodeproj-1.1.0/lib/xcodeproj/project/object/native_target.rb:462:in `block in add_file_references'
/usr/local/Cellar/cocoapods/1.0.1/libexec/gems/xcodeproj-1.1.0/lib/xcodeproj/project/object/native_target.rb:461:in `map'
/usr/local/Cellar/cocoapods/1.0.1/libexec/gems/xcodeproj-1.1.0/lib/xcodeproj/project/object/native_target.rb:461:in `add_file_references'
/usr/local/Cellar/cocoapods/1.0.1/libexec/gems/cocoapods-1.0.1/lib/cocoapods/installer/target_installer/pod_target_installer.rb:91:in `block (2 levels) in add_files_to_build_phases'
/usr/local/Cellar/cocoapods/1.0.1/libexec/gems/cocoapods-1.0.1/lib/cocoapods/installer/target_installer/pod_target_installer.rb:87:in `each'
/usr/local/Cellar/cocoapods/1.0.1/libexec/gems/cocoapods-1.0.1/lib/cocoapods/installer/target_installer/pod_target_installer.rb:87:in `block in add_files_to_build_phases'
/usr/local/Cellar/cocoapods/1.0.1/libexec/gems/cocoapods-1.0.1/lib/cocoapods/installer/target_installer/pod_target_installer.rb:76:in `each'
/usr/local/Cellar/cocoapods/1.0.1/libexec/gems/cocoapods-1.0.1/lib/cocoapods/installer/target_installer/pod_target_installer.rb:76:in `add_files_to_build_phases'
/usr/local/Cellar/cocoapods/1.0.1/libexec/gems/cocoapods-1.0.1/lib/cocoapods/installer/target_installer/pod_target_installer.rb:21:in `block in install!'
/usr/local/Cellar/cocoapods/1.0.1/libexec/gems/cocoapods-1.0.1/lib/cocoapods/user_interface.rb:141:in `message'
/usr/local/Cellar/cocoapods/1.0.1/libexec/gems/cocoapods-1.0.1/lib/cocoapods/installer/target_installer/pod_target_installer.rb:17:in `install!'
/usr/local/Cellar/cocoapods/1.0.1/libexec/gems/cocoapods-1.0.1/lib/cocoapods/installer.rb:621:in `block (2 levels) in install_libraries'
/usr/local/Cellar/cocoapods/1.0.1/libexec/gems/cocoapods-1.0.1/lib/cocoapods/installer.rb:619:in `each'
/usr/local/Cellar/cocoapods/1.0.1/libexec/gems/cocoapods-1.0.1/lib/cocoapods/installer.rb:619:in `block in install_libraries'
/usr/local/Cellar/cocoapods/1.0.1/libexec/gems/cocoapods-1.0.1/lib/cocoapods/user_interface.rb:141:in `message'
/usr/local/Cellar/cocoapods/1.0.1/libexec/gems/cocoapods-1.0.1/lib/cocoapods/installer.rb:618:in `install_libraries'
/usr/local/Cellar/cocoapods/1.0.1/libexec/gems/cocoapods-1.0.1/lib/cocoapods/installer.rb:170:in `block in generate_pods_project'
/usr/local/Cellar/cocoapods/1.0.1/libexec/gems/cocoapods-1.0.1/lib/cocoapods/user_interface.rb:63:in `section'
/usr/local/Cellar/cocoapods/1.0.1/libexec/gems/cocoapods-1.0.1/lib/cocoapods/installer.rb:167:in `generate_pods_project'
/usr/local/Cellar/cocoapods/1.0.1/libexec/gems/cocoapods-1.0.1/lib/cocoapods/installer.rb:119:in `install!'
/usr/local/Cellar/cocoapods/1.0.1/libexec/gems/cocoapods-1.0.1/lib/cocoapods/command/install.rb:37:in `run'
/Library/Ruby/Gems/2.0.0/gems/claide-1.0.0/lib/claide/command.rb:334:in `run'
/usr/local/Cellar/cocoapods/1.0.1/libexec/gems/cocoapods-1.0.1/lib/cocoapods/command.rb:50:in `run'
/usr/local/Cellar/cocoapods/1.0.1/libexec/gems/cocoapods-1.0.1/bin/pod:55:in `<top (required)>'
/usr/local/Cellar/cocoapods/1.0.1/libexec/bin/pod:23:in `load'
/usr/local/Cellar/cocoapods/1.0.1/libexec/bin/pod:23:in `<main>'
```

### PR DESCRIPTION
**Pull request by dnkoutso**
completes #5441 by adding specs
closes #5427 
closes #5525

### CODE REVIEW

**Comment by dnkoutso** [comment]
I am going to clean up the spec. its a bit crappy.

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb [hunk 1]
```diff
@@ -101,7 +101,11 @@ def add_files_to_build_phases
                 native_target.add_file_references(regular_file_refs, flags)
               end
 
-              header_file_refs = headers.map { |sf| project.reference_for_path(sf) }
+              header_file_refs = headers.map do |sf|
+                project.reference_for_path(sf).tap do |ref|
+                  raise "unable to find ref for #{sf}" unless ref
```
Comment (segiddins) [inline lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: should probably include the target as well, to make debugging easier?

**Comment by dnkoutso** [comment]
Updated spec. I think its cleaner now.

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb (hunk 1 above)
Comment (dnkoutso) [inline lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: Sure.

**Comment by dnkoutso** [comment]
Updated to include target name and spec to properly test the message output.

**Comment by dnkoutso** [comment]
@segiddins I just noticed this is incomplete and only handles invalid header file references but not source file references. I am updating it further.

**Comment by dnkoutso** [comment]
All set.

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb [hunk 2]
```diff
@@ -97,11 +97,19 @@ def add_files_to_build_phases
               }.each do |arc, files|
                 files = files - headers - other_source_files
                 flags = compiler_flags_for_consumer(consumer, arc)
-                regular_file_refs = files.map { |sf| project.reference_for_path(sf) }
+                regular_file_refs = files.map do |sf|
+                  project.reference_for_path(sf).tap do |ref|
+                    raise "unable to find source ref for #{sf} for target #{target.name}" unless ref
```
Comment (benasher44) [inline lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: These should probably be `raise Informative, '…`

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb [hunk 3]
```diff
@@ -97,11 +97,19 @@ def add_files_to_build_phases
...
+                  end
+                end
                 native_target.add_file_references(regular_file_refs, flags)
               end
 
-              header_file_refs = headers.map { |sf| project.reference_for_path(sf) }
+              header_file_refs = headers.map do |sf|
+                project.reference_for_path(sf).tap do |ref|
```
Comment (benasher44) [inline lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: This block looks about the same as above; you could probably wrap this into a private helper method that just takes the name of the kind of file for which you want to print the error ("header" or "source") as an arg.

**Comment by dnkoutso** [comment]
@benasher44 all done.

**Pull request merged**